content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,306 @@
+```
+
+Libraries like [Monocle][monocle] could be extended to reduce the boilerplate, but some boilerplate would remain.
+In addition, this breaks the intuitive similarity between construction and deconstruction.
+
+### Alternative desugaring
+
+As the above described desugaring has its drawbacks.Here are some alternatives with other drawbacks, and maybe better trade-offs.
+
+#### Use underscore methods
+
+In the sprit of [name based pattern matching](https://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html#name-based-match):
+
+```scala
+object User:
+  class UserMatching(user: User):
+    def _1 = user.name
+    def _name = user.name
+
+    @deprecatedName
+    def _oldName = user.name.toUpperCase
+    ...
+  
+  def unapply(user: User) = UserMatching(user)
+```
+
+Pro:
+
+* allows to have more named fields than positional fields
+* allows `@deprecatedName`
+
+Con:
+
+* How to detect that a name means the same as a position? Maybe detect simple patterns like the last line in the example?
+* It's long and verbose, without any shortcuts in sight.
+* An underscore at the beginning of a name isn't an unheard of pattern, even in Scala. This could accidentally expose fields, which weren't suppose to become fields.
+
+#### Annotated `unapply` method
+
+```scala
+object User:
+  @names("name", "age", "city")
+  def unapply(user: User): Some[(String, Age, String)] = ...
+```
+
+Pro:
+
+* simple to implement (was done in the first draft implementation)
+
+Con:
+
+* no clear way of encoding deprecated name
+* annotations are not very scalaish
+* no reuse possible
+
+#### Add type as vehicle
+
+```scala
+object User:
+  type Unapply = ("name", "age", "city")
+  def unapply(user: User): Some((String, Age, String)) = ...
+```
+
+Pro:
+
+* should be easy to implement
+* the type and the method can inherited from traits and allow some kind of reuse
+
+Con:
+
+* the type and the method can be defined in unrelated traits. Only at the use site of the can be checked if the type and the method agree on the arity of the pattern.
+* no clear way of encoding deprecated name
+
+#### Partial destructuring in guards
+
+Lionel Parreaux proposed a more powerful mechanism, where if guards of cases could them self contains destructuring patterns.
+
+```scala
+  case user: User if Age(years) <- user.age => years
+  case User(age = Age(years)) => years // both cases do the same thing
+```
+
+His proposal is strictly more powerful, but arguably less intuitive. Both, Pattern matching with named fields and Partial destructuring in guards could be implemented along each other. Named fields for simple patterns and destructuring in guards for complex patterns. However, they offer two ways to do the same thing and could lead to lots of bike shedding, if both got added to the language.
+
+## Open questions
+
+It could be useful to add extractors with just named fields to sealed traits and enums.
+
+What would reuse would look like? What is desirable?
```

Comment:
I don't understand what these mean.

Reply:
It was an idea in the back of my head, to add unapply methods to sealed trait with the names of the fields of the trait. I removed this section. Maybe the idea will come up again.